### PR TITLE
[IMP] web,sale: allow selection widget in kanban view

### PR DESCRIPTION
**Pull request by chevalierv**
Currently, users who want to change the visibility of product's
documents need to open their form view and edit it.

This new widget allows users to quickly edit the way a document is
displayed to their customers from the product's documents kanban view.

### CODE REVIEW

File: addons/sale/static/src/js/selection_kanban/selection_kanban_field.xml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<templates xml:space="preserve">
+
+    <t t-name="sale.SelectionKanbanField" t-inherit="web.SelectionField" t-inherit-mode="primary">
+        <xpath expr="//select" position="attributes">
```
Comment (Feyensv) [inline addons/sale/static/src/js/selection_kanban/selection_kanban_field.xml:5]:
```suggestion
        <select position="attributes">
```
?

File: addons/sale/static/src/js/selection_kanban/selection_kanban_field.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        switch (this.type) {
+            case "many2one":
+                if (value === false) {
+                    this.props.record.update({ [this.props.name]: false }, { save: this.props.autosave });
+                } else {
+                    this.props.record.update({
+                        [this.props.name]: this.options.find((option) => option[0] === value),
+                    }, { save: this.props.autosave });
+                }
+                break;
```
Comment (Feyensv) [inline addons/sale/static/src/js/selection_kanban/selection_kanban_field.js:28]: If the use is restricted to sale, should we really keep the support of m2o fields ?
Shouldn't we simply update the supportedTypes to restrict to selection fields, which is our only usecase for now?

Comment (chevalierv) [reply addons/sale/static/src/js/selection_kanban/selection_kanban_field.js:28]: Well, we don't have to. I restricted it to `selection` as you suggested

**Comment by chevalierv** [comment]
Hello @odoo/rd-framework-js :wave: 

Which solution do you prefer? 1694d1f7ee9117c27fd23b71edb021efb109958e or 3cbe12de2dffeaf43b24ee516f918d0a9f74b8e3. 

The second solution would allow `filterable_selection_field` to work with kanban whereas the first clearly separates the selection logic in kanban views from other views.

After your feedback about that, I'll adapt the tests and put them in `web`.

File: addons/web/static/src/views/fields/selection/selection_field.js [hunk 3]
```diff
@@ -86,14 +87,29 @@ export const selectionField = {
...
+    ],
     isEmpty: (record, fieldName) => record.data[fieldName] === false,
-    extractProps({ attrs }, dynamicInfo) {
+    extractProps({ attrs, options, viewType }, dynamicInfo) {
         return {
+            autosave: viewType === "kanban" && "autosave" in options ? Boolean(options.autosave) : true,
             placeholder: attrs.placeholder,
             required: dynamicInfo.required,
             domain: dynamicInfo.domain(),
+            readonly: dynamicInfo.readonly,
```
Comment (FrancoisGe) [inline addons/web/static/src/views/fields/selection/selection_field.js:109]: not a good idea, if i have a form view in readonly, my selection field shouldn't be in edit mode. Maybe ask to dbo, in kaban view, the selection shoud be in edit mode by default ?

Comment (chevalierv) [reply addons/web/static/src/views/fields/selection/selection_field.js:109]: cc @bouvyd

File: addons/web/static/src/views/fields/selection/selection_field.js [hunk 4]
```diff
@@ -13,6 +13,7 @@ export class SelectionField extends Component {
         placeholder: { type: String, optional: true },
         required: { type: Boolean, optional: true },
         domain: { type: Array, optional: true },
+        autosave: { type: Boolean, optional: true },
```
Comment (FrancoisGe) [inline addons/web/static/src/views/fields/selection/selection_field.js:16]: You should add tests in web/ for this props and a test with selection field in kanban

File: addons/website_sale/views/product_document_views.xml [hunk 5]
```diff
@@ -14,6 +14,20 @@
...
+        <field name="name">product.document.kanban.sale</field>
+        <field name="model">product.document</field>
+        <field name="inherit_id" ref="product.product_document_kanban"/>
+        <field name="arch" type="xml">
+            <div name="bottom" position="inside">
+                <div>
+                    <strong>Show on product page</strong>
+                    <field name="shown_on_product_page" class="ms-1" widget="boolean_toggle"/>
+                </div>
+            </div>
```
Comment (dalaOdoo) [inline addons/website_sale/views/product_document_views.xml:27]:
```suggestion
            <div name="bottom" position="inside">
                <div class="mt-2">
                    <span>Show on product page</span>
                    <field name="shown_on_product_page" class="ms-2" widget="boolean_toggle"/>
                </div>
            </div>
```

File: addons/sale/views/product_document_views.xml [hunk 6]
```diff
@@ -19,12 +19,12 @@
         <field name="model">product.document</field>
         <field name="inherit_id" ref="product.product_document_kanban"/>
         <field name="arch" type="xml">
-            <span name="bottom_left" position="inside">
-                <t t-if="record.attached_on.raw_value">
+            <div name="bottom" position="inside">
+                <div>
                     <strong>Visible at</strong>
```
Comment (dalaOdoo) [inline addons/sale/views/product_document_views.xml:24]:
```suggestion
                <div class="mt-2">
                    <span>Visible at</span>
```

File: addons/sale/views/product_document_views.xml [hunk 7]
```diff
@@ -19,12 +19,12 @@
...
-                    <field name="attached_on"/>
-                </t>
-            </span>
+                    <field name="attached_on" class="ms-1" widget="selection"/>
```
Comment (dalaOdoo) [inline addons/sale/views/product_document_views.xml:25]:
```suggestion
                    <field name="attached_on" class="ms-2" widget="selection"/>
```

File: addons/product/views/product_document_views.xml [hunk 8]
```diff
@@ -72,13 +72,7 @@
                                     <div class="o_kanban_record_body" name="body">
                                         <field name="url" widget="url" invisible="type == 'binary'"/>
```
Comment (dalaOdoo) [inline addons/product/views/product_document_views.xml:73]:
```suggestion
                                    <div class="o_kanban_record_body" name="body" t-att-class="'d-none' if type == 'binary' else ''" >
                                        <field name="url" widget="url"/>
```

File: addons/web/static/src/views/fields/selection/selection_field.js (hunk 3 above)
Comment (bouvyd) [reply addons/web/static/src/views/fields/selection/selection_field.js:109]: @FrancoisGe are you sure of what you're saying?

If I take the `boolean_toggle` as an example, we have something similar:
```js
export const booleanToggleField = {
    ...booleanField,
    component: BooleanToggleField,
    displayName: _t("Toggle"),
    supportedOptions: [
        {
            label: _t("Autosave"),
            name: "autosave",
            type: "boolean",
            default: true,
            help: _t(
                "If checked, the record will be saved immediately when the field is modified."
            ),
        },
    ],
    extractProps({ options }, dynamicInfo) {
        return {
            autosave: "autosave" in options ? Boolean(options.autosave) : true,
            readonly: dynamicInfo.readonly,
        };
    },
};
```
and yet in a form view with `edit="0"`, the widget is properly readonly.

Similarly in this branch, if I try that with a selection widget it is properly readonly.

Comment (FrancoisGe) [reply addons/web/static/src/views/fields/selection/selection_field.js:97]: @bouvyd Yes, you're right, I didn't give you the right use case.  In a form view, if you have an x2many in list view mode with a selection field. I think that all the selection fields will be in edit mode (if you have widget="selection" on the <field>). We have a BooleanToggleField extension for the list view.

File: addons/web/static/src/views/fields/selection/selection_field.js [hunk 9]
```diff
@@ -87,13 +88,16 @@ export const selectionField = {
...
     supportedTypes: ["many2one", "selection"],
     isEmpty: (record, fieldName) => record.data[fieldName] === false,
-    extractProps({ attrs }, dynamicInfo) {
+    extractProps({ attrs, viewType }, dynamicInfo) {
         return {
+            autosave: viewType === "kanban",
             placeholder: attrs.placeholder,
             required: dynamicInfo.required,
             domain: dynamicInfo.domain(),
+            readonly: dynamicInfo.readonly,
```
Comment (FrancoisGe) [inline addons/web/static/src/views/fields/selection/selection_field.js:97]:
```suggestion
        const props = {
            autosave: viewType === "kanban",
            placeholder: attrs.placeholder,
            required: dynamicInfo.required,
            domain: dynamicInfo.domain(),
            }
            if (viewType === "kanban") {
             props. readonly: dynamicInfo.readonly;
             }

        return props
```

File: addons/website_sale/views/product_document_views.xml [hunk 10]
```diff
@@ -14,6 +14,20 @@
         </field>
     </record>
 
+    <record id="product_document_kanban" model="ir.ui.view">
+        <field name="name">product.document.kanban.sale</field>
```
Comment (Feyensv) [inline addons/website_sale/views/product_document_views.xml:18]:
```suggestion
        <field name="name">product.document.kanban.website.sale</field>
```

File: addons/website_sale/views/product_document_views.xml [hunk 11]
```diff
@@ -14,6 +14,20 @@
...
+        <field name="model">product.document</field>
+        <field name="inherit_id" ref="product.product_document_kanban"/>
```
Comment (Feyensv) [inline addons/website_sale/views/product_document_views.xml:20]:
```suggestion
        <field name="inherit_id" ref="sale.product_document_kanban"/>
```
(personal pref, but not mandatory)

**Review by Feyensv — APPROVED** [review]
lgtm for sale/website_sale part ;)

**Comment by FrancoisGe** [comment]
@robodoo r+

**Comment by chevalierv** [comment]
@robodoo rebase-ff

**Pull request closed without merging**